### PR TITLE
Work on overloading by hardware target.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Whilst the build is green, I think `numba.tests.test_array_analysis.TestArrayAnalysis.test_stencilcall` is still broken with a lowering error but as stencil tests are python 3.7 only it could be on a non-py3.7 test slice.

File: numba/core/types/functions.py [hunk 1]
```diff
@@ -285,8 +285,43 @@ def get_call_type(self, context, args, kws):
...
+                   "registered. Given target was {}.")
+            raise ValueError(msg.format(target))
+
+        # fish out templates that are specific to the target if a target is
+        # specified
+        DEFAULT_HARDWARE = 'cpu'
+        usable = []
+        for ix, temp_cls in enumerate(self.templates):
+            # ? Need to do something about this next line
+            hw = temp_cls.metadata.get('hardware', DEFAULT_HARDWARE)
```
Comment (sklam) [inline numba/core/types/functions.py:311]: Maybe `DEFAULT_HARDWARE` should be generic?

File: numba/core/types/functions.py [hunk 2]
```diff
@@ -285,8 +285,43 @@ def get_call_type(self, context, args, kws):
...
+            target = hardware_registry.get(current_target(), None)
+
+        for k, v in hardware_registry.items():
+            if v == target:
+                target_hw = v
+                break
+        else:
+            msg = ("InternalError: The hardware target for TOS is not "
+                   "registered. Given target was {}.")
+            raise ValueError(msg.format(target))
```
Comment (sklam) [inline numba/core/types/functions.py:303]: These need to be encapsulated

File: numba/core/types/functions.py [hunk 3]
```diff
@@ -285,8 +285,43 @@ def get_call_type(self, context, args, kws):
...
+            hw = temp_cls.metadata.get('hardware', DEFAULT_HARDWARE)
+            if hw is not None:
+                hw_clazz = hardware_registry[hw]
+                if hw_clazz in target_hw.__mro__:
+                    usable.append((temp_cls, hw_clazz, ix))
+
+        # sort templates based on hardware specificity
+        def key(x):
+            return target_hw.__mro__.index(x[1])
+        order = [x[0] for x in sorted(usable, key=key)]
```
Comment (sklam) [inline numba/core/types/functions.py:320]: We can check for empty `order` here.

File: numba/core/types/functions.py (hunk 1 above)
Comment (sklam) [reply numba/core/types/functions.py:311]: Yes, that would get cuda to run

**Review by sklam — COMMENTED** [review]
I did a quick look and tried to get the CUDA target to work. The comments below show what I found.

The only pattern of CUDA test failures I found are related to "Intrinsic in function 'to_fixed_tuple". Some example of failed tests:

- `test_array_views (numba.cuda.tests.cudapy.test_cuda_array_interface.TestCudaArrayInterface)` 
- `test_strides (numba.cuda.tests.cudadrv.test_cuda_array_slicing.CudaArraySetting)`
- `test_rank (numba.cuda.tests.cudadrv.test_cuda_array_slicing.CudaArraySetting)`
- `test_broadcast (numba.cuda.tests.cudadrv.test_cuda_array_slicing.CudaArraySetting)`

My local diff to get cuda working:

```python
diff --git a/numba/core/types/functions.py b/numba/core/types/functions.py
index c3fb79337..a04343d17 100644
--- a/numba/core/types/functions.py
+++ b/numba/core/types/functions.py
@@ -304,7 +304,7 @@ class BaseFunction(Callable):
 
         # fish out templates that are specific to the target if a target is
         # specified
-        DEFAULT_HARDWARE = 'cpu'
+        DEFAULT_HARDWARE = 'generic'
         usable = []
         for ix, temp_cls in enumerate(self.templates):
             # ? Need to do something about this next line
diff --git a/numba/cuda/descriptor.py b/numba/cuda/descriptor.py
index 93fbaa083..3cd2a8f68 100644
--- a/numba/cuda/descriptor.py
+++ b/numba/cuda/descriptor.py
@@ -25,7 +25,7 @@ class CUDATarget(TargetDescriptor):
     @property
     def targetctx(self):
         if self._targetctx is None:
-            self._targetctx = CUDATargetContext(self._typingctx)
+            self._targetctx = CUDATargetContext(self._typingctx, target="cuda")
         return self._targetctx
```

**Comment by stuartarchibald** [comment]
> I did a quick look and tried to get the CUDA target to work. The comments below show what I found.
> 
> The only pattern of CUDA test failures I found are related to "Intrinsic in function 'to_fixed_tuple". Some example of failed tests:
> 
>     * `test_array_views (numba.cuda.tests.cudapy.test_cuda_array_interface.TestCudaArrayInterface)`
> 
>     * `test_strides (numba.cuda.tests.cudadrv.test_cuda_array_slicing.CudaArraySetting)`
> 
>     * `test_rank (numba.cuda.tests.cudadrv.test_cuda_array_slicing.CudaArraySetting)`
> 
>     * `test_broadcast (numba.cuda.tests.cudadrv.test_cuda_array_slicing.CudaArraySetting)`
> 
> 
> My local diff to get cuda working:

Thanks for this, applying the patch, I'm seeing the same/similar, all tests seem to fail at `to_fixed_tuple` on the CUDA target.

**Comment by stuartarchibald** [comment]
5dbeab7 gets CUDA working locally.

**Comment by stuartarchibald** [comment]
4d44673 Adds tests based on the gist at the top of https://github.com/numba/numba/pull/6598

File: numba/core/typing/templates.py [hunk 4]
```diff
@@ -720,7 +724,56 @@ def _build_impl(self, cache_key, args, kws):
...
+                    msg = ("Unknown hardware target '{}', has it been ",
+                           "registered?")
+                    report_unknown_hardware(msg)
+
+                # scan for the the current hardware target it's stuffed into the
+                # call stack at present, get that.
+                if len(self.context.callstack._stack) > 0:
+                    tos_target = self.context.callstack[0].target
+                else:
+                    tos_target = hardware_registry.get(current_target(), None)
```
Comment (stuartarchibald) [inline numba/core/typing/templates.py:756]: @sklam this is where the stack issues we discussed OOB stem from. The issue is essentially that if you enter the compilation chain from somewhere less standard then there could well be no frames and therefore no stack. This sort of thing being an example of said problem: https://github.com/numba/numba/pull/6898/files#diff-0e3c8738578e42578741958d965937324c13aaf8e6fc1523ce6affd318013dc3R592-R603

File: numba/core/lowering.py [hunk 5]
```diff
@@ -983,7 +983,15 @@ def _lower_call_normal(self, fnty, expr, signature):
             argvals = self.fold_call_args(
                 fnty, signature, expr.args, expr.vararg, expr.kws,
             )
-        impl = self.context.get_function(fnty, signature)
+        tname = expr.hardware
+        if tname is not None:
+            from numba.core.extending_hardware import (dispatcher_registry,
+                                                       hardware_registry)
+            disp = dispatcher_registry[hardware_registry[tname]]
```
Comment (sklam) [inline numba/core/lowering.py:990]: The pattern of `dispatcher_registry[hardware_registry[tname]]` has appeared multiple times. It's common enough to have a method for it.

File: numba/core/extending.py [hunk 6]
```diff
@@ -57,7 +57,7 @@ def generic(self):
 
 
 def overload(func, jit_options={}, strict=True, inline='never',
-             prefer_literal=False):
+             prefer_literal=False, **kwargs):
```
Comment (sklam) [inline numba/core/extending.py:60]: Can we leave a note about `hardware` being the only accepted key for kwargs?

File: numba/core/typing/templates.py [hunk 7]
```diff
@@ -720,7 +724,56 @@ def _build_impl(self, cache_key, args, kws):
             On failure, returns `(None, None)`.
 
         """
+        from numba.core import decorators
         from numba import jit
+        jitter_str = self.metadata.get('hardware', None)
```
Comment (sklam) [inline numba/core/typing/templates.py:729]: We should move this entire `jitter` selection logic out into its own method

File: numba/core/extending_hardware.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
+from abc import ABC, abstractmethod
```
Comment (sklam) [inline numba/core/extending_hardware.py:1]: In the next PR, we should make sure this file can be imported in the top everywhere.

**Comment by sklam** [comment]
There's an empty `hardware.py` file that we probably be removed.

File: numba/core/extending_hardware.py (hunk 8 above)
Comment (stuartarchibald) [reply numba/core/extending_hardware.py:1]: Agree.

File: numba/core/typing/templates.py (hunk 7 above)
Comment (stuartarchibald) [reply numba/core/typing/templates.py:706]: Agree, done.

File: numba/core/extending.py (hunk 6 above)
Comment (stuartarchibald) [reply numba/core/extending.py:60]: Agree, done.

File: numba/core/lowering.py (hunk 5 above)
Comment (stuartarchibald) [reply numba/core/lowering.py]: Yes! Done.

File: numba/core/types/functions.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/core/types/functions.py]: done.

File: numba/core/types/functions.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/core/types/functions.py:308]: done, and removed the later check if there's zero failures as I don't think it's reachable if caught here.

File: numba/core/types/functions.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/core/types/functions.py]: done

**Comment by stuartarchibald** [comment]
> There's an empty `hardware.py` file that we probably be removed.

Done in 9166c3f

**Comment by stuartarchibald** [comment]
Testing CPU and CUDA with build jobs:

Buildfarm ID: `numba_smoketest_cpu_yaml_14`.
Buildfarm ID: `numba_smoketest_cuda_yaml_42`.

**Comment by stuartarchibald** [comment]
> Buildfarm ID: `numba_smoketest_cuda_yaml_42`.

CUDA passed.

**Review by sklam — APPROVED** [review]
From OOB discussion, we want to move forward quickly and get this PR in since there are other PRs depending of features added here. I'm happy to merge this as soon as the buildfarm passes

**Comment by stuartarchibald** [comment]
There's a load of tests (seemingly randomly, i.e. no pattern across hardware or pythons) failing. The only thing I can see in common is that they are tests which rely on functions jitted on initialisation, e.g. dictionary support and jitclass support.

**Comment by stuartarchibald** [comment]
> Buildfarm ID: numba_smoketest_cpu_yaml_14.

Failed with https://github.com/numba/numba/pull/6898#issuecomment-819664405, I think 0ef4c1b may fix.

**Comment by stuartarchibald** [comment]
Testing CPU and CUDA with build jobs for 0ef4c1b:

Buildfarm ID: `numba_smoketest_cpu_yaml_15`.
Buildfarm ID: `numba_smoketest_cuda_yaml_43`.

**Comment by stuartarchibald** [comment]
> Testing CPU and CUDA with build jobs for [0ef4c1b](https://github.com/numba/numba/commit/0ef4c1b6b8c5ecf27493966b11f83b7a4be3afbb):
> 
> Buildfarm ID: `numba_smoketest_cpu_yaml_15`.

is ok

> Buildfarm ID: `numba_smoketest_cuda_yaml_43`.

is broken, CUDA can't find tuple setitem intrinsics, most likely due to the builtins registry not being used as the place to register the impl from the CPU side.

**Comment by stuartarchibald** [comment]
Testing CPU and CUDA with build jobs for 913ac82:

Buildfarm ID: `numba_smoketest_cpu_yaml_16`.
Buildfarm ID: `numba_smoketest_cuda_yaml_45`.

**Comment by stuartarchibald** [comment]
> Testing CPU and CUDA with build jobs for [913ac82](https://github.com/numba/numba/commit/913ac82f17cdcf8b7319745d4ffa1169972dbd1b):
> 
> Buildfarm ID: `numba_smoketest_cpu_yaml_16`.

Waiting on Aarch64 builds but otherwise fine on all builds.

> Buildfarm ID: `numba_smoketest_cuda_yaml_45`.

Passed.

**Review by sklam — APPROVED** [review]
I can understand the latest commits to fix the CI failures.

**Pull request merged**